File: mcp_observatory/proposal_commit/storage.py
"""Storage adapters for proposal/commit records and nonce replay protection."""

from __future__ import annotations

import json
import os
from abc import ABC, abstractmethod
from datetime import datetime, timezone
from typing import Any, Optional

import asyncpg


def utc_now() -> datetime:
    return datetime.now(timezone.utc)


class ProposalCommitStorage(ABC):
    """Abstract storage backend for proposals, commits and nonce replay checks."""

    @abstractmethod
    async def get_baseline_prompt_hash(self, tool_name: str) -> Optional[str]:
        """Fetch baseline prompt hash configured for a tool."""

    @abstractmethod
    async def set_baseline_prompt_hash(self, tool_name: str, prompt_hash: str) -> None:
        """Set baseline prompt hash for a tool."""

    @abstractmethod
    async def save_proposal(
        self,
        *,
        proposal_id: str,
        tool_name: str,
        args_json: str,
        prompt_hash: str,
        composite_score: float,
        decision: str,
        created_at: datetime,
    ) -> None:
        """Persist proposal row."""

    @abstractmethod
    async def get_proposal(self, proposal_id: str) -> Optional[dict[str, Any]]:
        """Fetch proposal by ID."""

    @abstractmethod
    async def save_commit(
        self,
        *,
        commit_id: str,
        proposal_id: str,
        token_id: Optional[str],
        decision: str,
        verification_reason: str,
        created_at: datetime,
    ) -> None:
        """Persist commit row."""

    @abstractmethod
    async def nonce_seen(self, nonce: str, token_id: str, expires_at: datetime) -> bool:
        """Return True if nonce was already seen and active; otherwise store it and return False."""


class InMemoryStorage(ProposalCommitStorage):
    """In-memory storage fallback backend."""

    def __init__(self) -> None:
        self.baseline: dict[str, str] = {}
        self.proposals: dict[str, dict[str, Any]] = {}
        self.commits: dict[str, dict[str, Any]] = {}
        self.nonces: dict[str, tuple[str, datetime]] = {}

    async def get_baseline_prompt_hash(self, tool_name: str) -> Optional[str]:
        return self.baseline.get(tool_name)

    async def set_baseline_prompt_hash(self, tool_name: str, prompt_hash: str) -> None:
        self.baseline[tool_name] = prompt_hash

    async def save_proposal(self, **kwargs: Any) -> None:
        self.proposals[kwargs["proposal_id"]] = dict(kwargs)

    async def get_proposal(self, proposal_id: str) -> Optional[dict[str, Any]]:
        return self.proposals.get(proposal_id)

    async def save_commit(self, **kwargs: Any) -> None:
        self.commits[kwargs["commit_id"]] = dict(kwargs)

    async def nonce_seen(self, nonce: str, token_id: str, expires_at: datetime) -> bool:
        now = utc_now()
        expired = [k for k, (_, exp) in self.nonces.items() if exp <= now]
        for key in expired:
            self.nonces.pop(key, None)

        if nonce in self.nonces:
            return True
        self.nonces[nonce] = (token_id, expires_at)
        return False


class PostgresStorage(ProposalCommitStorage):
    """Postgres-backed storage using asyncpg."""

    def __init__(self, dsn: str) -> None:
        self.dsn = dsn
        self.pool: asyncpg.Pool | None = None

    async def connect(self) -> None:
        if self.pool is None:
            self.pool = await asyncpg.create_pool(dsn=self.dsn, min_size=1, max_size=4)

    async def close(self) -> None:
        if self.pool is not None:
            await self.pool.close()
            self.pool = None

    async def get_baseline_prompt_hash(self, tool_name: str) -> Optional[str]:
        await self.connect()
        assert self.pool is not None
        async with self.pool.acquire() as conn:
            row = await conn.fetchrow("SELECT prompt_hash FROM tool_prompt_baselines WHERE tool_name=$1", tool_name)
            return row["prompt_hash"] if row else None

    async def set_baseline_prompt_hash(self, tool_name: str, prompt_hash: str) -> None:
        await self.connect()
        assert self.pool is not None
        async with self.pool.acquire() as conn:
            await conn.execute(
                """
                INSERT INTO tool_prompt_baselines (tool_name, prompt_hash)
                VALUES ($1, $2)
                ON CONFLICT (tool_name) DO UPDATE SET prompt_hash = EXCLUDED.prompt_hash
                """,
                tool_name,
                prompt_hash,
            )

    async def save_proposal(self, **kwargs: Any) -> None:
        await self.connect()
        assert self.pool is not None
        async with self.pool.acquire() as conn:
            await conn.execute(
                """
                INSERT INTO proposals (proposal_id, tool_name, args_json, prompt_hash, composite_score, decision, created_at)
                VALUES ($1, $2, $3::jsonb, $4, $5, $6, $7)
                """,
                kwargs["proposal_id"],
                kwargs["tool_name"],
                kwargs["args_json"],
                kwargs["prompt_hash"],
                kwargs["composite_score"],
                kwargs["decision"],
                kwargs["created_at"],
            )

    async def get_proposal(self, proposal_id: str) -> Optional[dict[str, Any]]:
        await self.connect()
        assert self.pool is not None
        async with self.pool.acquire() as conn:
            row = await conn.fetchrow("SELECT * FROM proposals WHERE proposal_id=$1", proposal_id)
            if row is None:
                return None
            return dict(row)

    async def save_commit(self, **kwargs: Any) -> None:
        await self.connect()
        assert self.pool is not None
        async with self.pool.acquire() as conn:
            await conn.execute(
                """
                INSERT INTO commits (commit_id, proposal_id, token_id, decision, verification_reason, created_at)
                VALUES ($1, $2, $3, $4, $5, $6)
                """,
                kwargs["commit_id"],
                kwargs["proposal_id"],
                kwargs["token_id"],
                kwargs["decision"],
                kwargs["verification_reason"],
                kwargs["created_at"],
            )

    async def nonce_seen(self, nonce: str, token_id: str, expires_at: datetime) -> bool:
        await self.connect()
        assert self.pool is not None
        async with self.pool.acquire() as conn:
            async with conn.transaction():
                await conn.execute("DELETE FROM nonces WHERE expires_at <= NOW()")
                exists = await conn.fetchrow("SELECT nonce FROM nonces WHERE nonce=$1", nonce)
                if exists:
                    return True
                await conn.execute(
                    "INSERT INTO nonces (nonce, token_id, expires_at) VALUES ($1, $2, $3)",
                    nonce,
                    token_id,
                    expires_at,
                )
                return False


def create_storage_from_env() -> ProposalCommitStorage:
    """Create Postgres storage if env configured, otherwise in-memory."""
    dsn = os.getenv("MCP_OBSERVATORY_PG_DSN") or os.getenv("DATABASE_URL")
    if dsn:
        return PostgresStorage(dsn=dsn)
    return InMemoryStorage()

File: src/proposal/storage.ts
import { Pool } from 'pg';

export interface Proposal {
  id: string;
  toolName: string;
  toolArgsHash: string;
  status: 'allowed' | 'blocked' | 'review';
  createdAt: Date;
  expiresAt: Date;
}

export interface Commit {
  id: string;
  proposalId: string;
  executed: boolean;
  executedAt?: Date;
}

export class InMemoryProposalStorage {
  private proposals: Map<string, Proposal> = new Map();
  private commits: Map<string, Commit> = new Map();

  async saveProposal(proposal: Proposal): Promise<void> {
    this.proposals.set(proposal.id, proposal);
  }

  async getProposal(id: string): Promise<Proposal | null> {
    const proposal = this.proposals.get(id);
    if (!proposal) return null;

    if (proposal.expiresAt < new Date()) {
      this.proposals.delete(id);
      return null;
    }

    return proposal;
  }

  async saveCommit(commit: Commit): Promise<void> {
    this.commits.set(commit.id, commit);
  }

  async getCommit(id: string): Promise<Commit | null> {
    return this.commits.get(id) || null;
  }

  async saveNonce(nonce: string): Promise<void> {
    // In-memory storage for nonce tracking
  }

  async hasNonce(nonce: string): Promise<boolean> {
    // Would need separate nonce tracking
    return false;
  }

  async cleanup(): Promise<void> {
    const now = new Date();
    for (const [id, proposal] of this.proposals.entries()) {
      if (proposal.expiresAt < now) {
        this.proposals.delete(id);
      }
    }
  }
}

export class PostgresProposalStorage {
  private pool: Pool;

  constructor(dsn: string) {
    this.pool = new Pool({ connectionString: dsn });
  }

  async saveProposal(proposal: Proposal): Promise<void> {
    const query = `
      INSERT INTO proposals (id, tool_name, tool_args_hash, status, created_at, expires_at)
      VALUES ($1, $2, $3, $4, $5, $6)
      ON CONFLICT (id) DO UPDATE SET status = $4, expires_at = $6
    `;
    await this.pool.query(query, [
      proposal.id,
      proposal.toolName,
      proposal.toolArgsHash,
      proposal.status,
      proposal.createdAt,
      proposal.expiresAt,
    ]);
  }

  async getProposal(id: string): Promise<Proposal | null> {
    const result = await this.pool.query(
      'SELECT * FROM proposals WHERE id = $1 AND expires_at > NOW()',
      [id]
    );
    if (result.rows.length === 0) return null;
    const row = result.rows[0];
    return {
      id: row.id,
      toolName: row.tool_name,
      toolArgsHash: row.tool_args_hash,
      status: row.status,
      createdAt: row.created_at,
      expiresAt: row.expires_at,
    };
  }

  async saveCommit(commit: Commit): Promise<void> {
    const query = `
      INSERT INTO commits (id, proposal_id, executed, executed_at)
      VALUES ($1, $2, $3, $4)
    `;
    await this.pool.query(query, [
      commit.id,
      commit.proposalId,
      commit.executed,
      commit.executedAt,
    ]);
  }

  async getCommit(id: string): Promise<Commit | null> {
    const result = await this.pool.query('SELECT * FROM commits WHERE id = $1', [id]);
    if (result.rows.length === 0) return null;
    const row = result.rows[0];
    return {
      id: row.id,
      proposalId: row.proposal_id,
      executed: row.executed,
      executedAt: row.executed_at,
    };
  }

  async saveNonce(nonce: string): Promise<void> {
    await this.pool.query(
      'INSERT INTO nonces (nonce, created_at) VALUES ($1, NOW())',
      [nonce]
    );
  }

  async hasNonce(nonce: string): Promise<boolean> {
    const result = await this.pool.query('SELECT 1 FROM nonces WHERE nonce = $1', [
      nonce,
    ]);
    return result.rows.length > 0;
  }

  async cleanup(): Promise<void> {
    await this.pool.query('DELETE FROM proposals WHERE expires_at < NOW()');
  }

  async close(): Promise<void> {
    await this.pool.end();
  }
}
